pkg/runtime/bulk_subscriber.go

```diff
@@ -259,80 +266,98 @@ func (a *DaprRuntime) createEnvelopeAndInvokeSubscriber(ctx context.Context, psm
 	if err != nil {
 		id = idObj.String()
 	}
-	psm.cloudEvents = psm.cloudEvents[:psm.length]
-	psm.rawData = psm.rawData[:psm.length]
-	psm.entries = psm.entries[:psm.length]
+	psm.pubSubMessages = psm.pubSubMessages[:psm.length]
+	psm.path = path
 	envelope := runtimePubsub.NewBulkSubscribeEnvelope(&runtimePubsub.BulkSubscribeEnvelope{
 		ID:       id,
 		Topic:    topic,
-		Entries:  psm.rawData,
 		Pubsub:   psName,
 		Metadata: msg.Metadata,
 	})
+	var firstAttempt atomic.Bool
+	firstAttempt.Store(true)
+	_, err = policy(func(ctx context.Context) (any, error) {
+		if !firstAttempt.Load() {
+			newPubSubMsg := utils.Filter(psm.pubSubMessages, func(ps pubSubMessage) bool {
+				index := (*entryIdIndexMap)[ps.entry.EntryId]
+				return (*bulkResponses)[index].Error != nil
+			}, firstAttempt.Load())
+			psm.pubSubMessages = newPubSubMsg
+			psm.length = len(newPubSubMsg)
+		} else {
+			firstAttempt.CompareAndSwap(true, false)
```

Comment:
This can be simplified to a `Store` instead of `CompareAndSwap`, right?

Reply:
CAS is correct to use in this case to ensure the operation is actually atomic, but it shouldn't be done here because this can lead to race conditions: while operations like `Load` are performed atomically, there's no lock around them. CAS returns a boolean indicating if the swap happened and you can use that instead:

```go
// If the swap happened, returns true
if firstAttempt.CompareAndSwap(true, false) {
	newPubSubMsg := utils.Filter(psm.pubSubMessages, func(ps pubSubMessage) bool {
		index := (*entryIdIndexMap)[ps.entry.EntryId]
		return (*bulkResponses)[index].Error != nil
	}, false) // We can put false directly because we know it's false in this block
	psm.pubSubMessages = newPubSubMsg
	psm.length = len(newPubSubMsg)
}
```

Reply:
Actually, I had tried to look in that direction when your PR https://github.com/dapr/dapr/pull/5538 was raised, but issue is that in case of these Bulk APIs, we want to keep trying (say retries) decrementally i.e. omitting out all Successful entries and just want to keep re-try for Failed entries. And, that all becomes a part of Policy Only, But if we return from policy, we get outside Policy. 
How can we refactor so that we use return of `policy` but also remain inside it?

Reply:
Ok, I see what you're trying to do now. I don't think this can be done without side effects, indeed. You should be able to make this work safely, even with side effects, by relying on atomic's and CAS. Ideally you will want to use [`atomic.Pointer`](https://pkg.go.dev/sync/atomic#Pointer) to keep a pointer to the last list of objects that need to be requested It's going to be a bit tricky though :)

Reply:
As discussed with @ItalyPaleAle and @mukundansundar, changed the implementation. Using Accumulator now from NewRunnerWithOptions.